src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -0,0 +1,195 @@
+        byte[] encodedParams = use(EvpKey::getDerEncodedParams);
+        try {
+            AlgorithmParameters params = AlgorithmParameters.getInstance(type.jceName);
+            params.init(encodedParams);
+            return params.getParameterSpec(paramSpec);
+        } catch (final GeneralSecurityException | IOException ex) {
+            throw new RuntimeCryptoException("Unable to deserialize parameters: "
+                + Base64.getEncoder().encodeToString(encoded), ex);
+        }
+    }
+
+    /**
+     * This method will be called by @{link #destroy()} after calling @{code internalKey.release()}.
+     */
+    protected synchronized void destroyJavaState() {
+        // NOP
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        // We try to avoid comparing the encoded values
+        // because it may be slow and may pull secret data into the Java heap
+        if (this == obj) {
+            return true;
+        }
+        if (!(obj instanceof Key)) { // Implicit null check
+            return false;
+        }
+        final Key other = (Key) obj;
+        if (!getAlgorithm().equalsIgnoreCase(other.getAlgorithm())) {
+            return false;
+        }
+
+        final byte[] otherEncoded;
+        if (obj.getClass().equals(getClass())) {
+            // If it is also an EvpKey then we can see if the internal key is the same
+            EvpKey evpOther = (EvpKey) obj;
+            if (internalKey.equals(evpOther.internalKey)) {
+                return true;
+            }
+            // If it is also an EvpKey then we can grab the other encoded value without copying it
+            evpOther.initEncoded();
+            otherEncoded = evpOther.encoded;
+        } else {
+            otherEncoded = other.getEncoded();
+        }
+
+        // Constant time equality check
+        initEncoded();
+        return MessageDigest.isEqual(encoded, otherEncoded);
+    }
+
+    @Override
+    public int hashCode() {
+        // TODO: Consider ways to avoid exposing the entire encoded object ot Java for private keys just for a hashCode
+        if (cachedHashCode == null) {
+            synchronized (this) {
+                initEncoded();
+                // Selected to match implementations of sun.security.pkcs.PKCS8Key and sun.security.x509.X509Key
+                int workingValue = 0;
+                if (isPublicKey) {
+                    workingValue = encoded.length;
+                    for (final byte b : encoded) {
+                        workingValue += (b & 0xff) * 37;
+                    }
+                } else {
+                    if (Utils.getJavaVersion() >= 17) {
+                        workingValue = Arrays.hashCode(encoded);
+                    } else {
+                        for (int idx = 0; idx < encoded.length; idx++) {
+                            workingValue += encoded[idx] * idx;
+                        }
```

Comment:
Do we have a test to confirm that the hash code produced here matched the implementation of sun.security.x509.X509Key?  It looks different than the result I found when I googled for it: https://github.com/frohoff/jdk8u-jdk/blob/master/src/share/classes/sun/security/x509/X509Key.java#L458

```
    public int hashCode() {
        try {
            byte[] b1 = getEncodedInternal();
            int r = b1.length;
            for (int i = 0; i < b1.length; i++) {
                r += (b1[i] & 0xff) * 37;
            }
            return r;
           ...
```